lib/oauth2-client.js

```diff
@@ -0,0 +1,185 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import database from 'bedrock-mongodb';
+import * as bedrock from 'bedrock';
+import {httpClient} from '@digitalbazaar/http-client';
+import logger from './logger.js';
+const {util: {BedrockError}} = bedrock;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await database.openCollections(['access-tokens']);
+
+  await database.createIndexes([{
+    // prevent duplicate tokens from being stored
+    collection: 'access-tokens',
+    fields: {'token.token': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire tokens with an `expires` date field
+    collection: 'access-tokens',
+    fields: {'token.expires': 1},
+    options: {
+      partialFilterExpression: {'token.expires': {$exists: true}},
+      unique: false,
+      background: false,
+      expireAfterSeconds: 0
+    }
+  }]);
+});
+
+export class OAuth2Client {
+/**
+ * Constructs a OAuth2Client instance that can be used to return an accessToken.
+ *  It first checks the database to see if an accessToken already exists for the
+ *  client id; if not it creates an accessToken then stores it in the database.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.config - The "oauth2-client-grant" properties.
+ * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+ *
+ * @returns {OAuth2Client} The `OAuth2Client` instance.
+ */
+  constructor({config, httpsAgent} = {}) {
+    this.client_id = config.client_id;
+    this.client_secret = config.client_secret;
+    this.token_endpoint = config.token_endpoint;
+    this.grant_type = config.grant_type;
+    this.scope = config.scope;
+    this.maxRetries = config.maxRetries;
+    this.httpsAgent = httpsAgent;
+  }
+  /**
+   * Gets a new accessToken by first checking the database to see if an
+   *  accessToken already exists for the client id; if not it creates an
+   *  accessToken then stores it in the database.
+   *
+   * @returns {string} The accessToken.
+   */
+  async getToken() {
```

Comment:
Now that I see that we have a different method that actually gets a token from the database, this should be changed. What I'd recommend is moving all of the database functions into a separate file: `storage.js` and making them use the naming convention `insert()`, `get()`, etc. Then the main file can continue using the name that's being used here. That way it's clear when you're calling `storage.get()` that you're hitting storage to get something. That's the model we use in a number of other places.

Reply:
This pattern would also make things easier to review and maintain as it has a clear separation of concerns via different files.